service/src/ServiceIOGroup.cpp

```diff
@@ -155,42 +170,93 @@ namespace geopm
+        request.domain = domain_type;
+        request.domain_idx = domain_idx;
+        request.name[NAME_MAX - 1] = '\0';
+        strncpy(request.name, signal_name_strip.c_str(), NAME_MAX - 1);
+        m_signal_requests.push_back(request);
+        return m_signal_requests.size() - 1;
     }
 
     int ServiceIOGroup::push_control(const std::string &control_name,
                                      int domain_type,
                                      int domain_idx)
     {
-        throw Exception("ServiceIOGroup::push_control()",
-                        GEOPM_ERROR_NOT_IMPLEMENTED, __FILE__, __LINE__);
-        return -1;
+        if (!is_valid_control(control_name)) {
+            throw Exception("ServiceIOGroup::push_control(): control name \"" +
+                            control_name + "\" not found",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_type != control_domain_type(control_name)) {
+            throw Exception("ServiceIOGroup::push_control(): domain_type requested does not match the domain of the control (" + control_name + ").",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_idx < 0 || domain_idx >= m_platform_topo.num_domain(domain_type)) {
+            throw Exception("ServiceIOGroup::push_control(): domain_idx out of range",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        std::string control_name_strip = strip_plugin_name(control_name);
+        if (control_name_strip.size() >= NAME_MAX) {
+            throw Exception("ServiceIOGroup::push_control(): control_name: " + control_name + " is too long",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        geopm_request_s request;
+        request.domain = domain_type;
+        request.domain_idx = domain_idx;
+        request.name[NAME_MAX - 1] = '\0';
+        strncpy(request.name, control_name_strip.c_str(), NAME_MAX - 1);
+        m_control_requests.push_back(request);
+        return m_control_requests.size() - 1;
     }
 
     void ServiceIOGroup::read_batch(void)
     {
-        // No implementation until we support push_signal()
+        init_batch_server();
+        if (m_is_batch_active &&
+            m_signal_requests.size() != 0) {
+            m_batch_samples = m_batch_client->read_batch();
+        }
     }
 
     void ServiceIOGroup::write_batch(void)
     {
-        // No implementation until we support push_control()
+        if (m_is_batch_active &&
+            m_control_requests.size() != 0) {
+            m_batch_client->write_batch(m_batch_settings);
+        }
     }
 
     double ServiceIOGroup::sample(int sample_idx)
     {
-        throw Exception("ServiceIOGroup::sample()",
-                        GEOPM_ERROR_NOT_IMPLEMENTED, __FILE__, __LINE__);
-        return NAN;
+        // TODO: Check that sample_idx is in range and batch_server is not null
```

Comment:
TODO for this change, or future changes?

Reply:
hmmm....Yeah that seems like it should be DONE not TODO...